### PR TITLE
Bcrypt Hash Authentication p1

### PR DESCRIPTION
**Pull request by gdunstone**
web services changes required for secure hash authentication see: http://php.net/manual/en/faq.passwords.php

I added a dependency for `at.favre.lib.crypto.bcrypt.BCrypt`. I'm open to other suggestions for libraries however this one was straightforward and required minimal boilerplate.

This pull request depends on https://github.com/OpenSILEX/phis-webapp/pull/101

### CODE REVIEW

**Comment by gdunstone** [comment]
This will require users to reset their passwords.

**Comment by vincentmigot** [comment]
Hello Gareth,
Pierre-Etienne and I have been looking at your pull request, thank you for it !
We agree with you for the need of moving password encryption from md5 to bcrypt but since that will impact all users we need a green light from our supervisor first.

On a more technical level could we ask you to do some changes on your pull request ?
- On the web service project, please remove your docker script and configuration files
- On the web application project, please keep only the modification on SiteController.php and UserController.php

If you need any other information please feel free to contact us on slack, we can even plan for a web conference if necessary.

**Comment by gdunstone** [comment]
I removed the dockerization.
however I left some of the configuration file changes for the webapp as they appear to be leftover from an earlier version and it doesn't work without those changes.

**Comment by vincentmigot** [comment]
Hello and thank you for your modifications.
After discussion, as we can not actually force our users to reset their passwords on some platforms, we think it's better to offer the choice of password encryption based on some config parameters with a solution like :
- In the webapp
Add a parameter called something like passwordEncryption in config/params.php which can have MD5 or BCRYPT value and use this parameter to determine the password encryption method.

- In the webservice
Add the same parameter as in the webapp in src/main/resources/service.properties and do password check depending on it.

If that solution suits you, we will be pleased to integrate it.

**Comment by zarlo** [comment]
@vincentmigot what if we made it automatically update the md5 passwords to bcrypt

**Comment by vincentmigot** [comment]
Hi, 
Of course if you have a reliable way to "unhash" md5 passwords it could be the best solution.
If you can give us a script which can do that, we can integrate it on our already deployed instance.

**Comment by RichardPoire** [comment]
> Hi,
> Of course if you have a reliable way to "unhash" md5 passwords it could be the best solution.
> If you can give us a script which can do that, we can integrate it on our already deployed instance.

Hi, 
it is inherently impossible to "unhash" a password reliably. It is actually the purpose of hashing in the first place to maintain data integrity.

I am not quite clear to why it is not be possible to force a password reset on some platforms.
Now that this rather critical issue is spotted, wouldn't it be easier to fix it now than later when it will impact more users?

File: phis2-ws/.classpath [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vincentmigot) [inline phis2-ws/.classpath]: Please remove this file

File: phis2-ws/.project [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vincentmigot) [inline phis2-ws/.project]: Please remove this file

File: phis2-ws/.settings/org.eclipse.core.resources.prefs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+eclipse.preferences.version=1
```
Comment (vincentmigot) [inline phis2-ws/.settings/org.eclipse.core.resources.prefs]: Please remove this file

File: phis2-ws/.settings/org.eclipse.jdt.apt.core.prefs [hunk 4]
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
Comment (vincentmigot) [inline phis2-ws/.settings/org.eclipse.jdt.apt.core.prefs]: Please remove this file

File: phis2-ws/.settings/org.eclipse.jdt.core.prefs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+eclipse.preferences.version=1
```
Comment (vincentmigot) [inline phis2-ws/.settings/org.eclipse.jdt.core.prefs]: Please remove this file

File: phis2-ws/.settings/org.eclipse.m2e.core.prefs [hunk 6]
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
```
Comment (vincentmigot) [inline phis2-ws/.settings/org.eclipse.m2e.core.prefs]: Please remove this file

File: phis2-ws/src/main/java/phis2ws/service/resources/brapi/TokenResourceService.java [hunk 7]
```diff
@@ -336,28 +339,44 @@ private String createId(String username) throws NoSuchAlgorithmException, SQLExc
...
+
                     uspb.admin = uspb.isAdmin(user);
+
                 } else {
-                    user = null;
+                    // MD5 hash fallback
+                    if (password.equals(user.getPassword())) {
+                        uspb.admin = uspb.isAdmin(user);
+                    }
+                    else
```
Comment (vincentmigot) [inline phis2-ws/src/main/java/phis2ws/service/resources/brapi/TokenResourceService.java]: Please remove unnecessary blank lines and leave brackets on the same line as "else" keyword

**Comment by vincentmigot** [comment]
Hi, 
This solution is fine to us.
I do some review comments on your pull requests about our coding style and some unnecessary files to remove.
Once it's done we will integrate it.

**Comment by vincentmigot** [comment]
Hi, 

Sorry not having responding you for a long time.
I tried today to integrate your pull request and actually test it.
So we can't integrate it right now because it's not working due to fact that BCrypt generate a new different hash for the same string each time you call the "password_hash" function.
You can always validate each hash with the original string but you can't validate hashes of the same string against each other and that's what you are trying to do in the webservice.

We could imagine some workarounds to make it work and we can talk about it on our Slack thread if you liked to.

But in order to be transparent with you, we plan to have a big rework on how users are managed in our software improving security and integration with other information systems :
- by getting rid of md5
- by allowing connexion against existing system with Oauth, LDAP, ...
- by sharing users between multiple OpenSilex instances

If you have other needs we will be pleased to take them in account during our rework or help you to work with us on them.

For now I will close your pull request, don't hesitate to contact us if you wan't more explaination.

Vincent

**Pull request closed without merging**